src/lib/BridgeBase.sol

```diff
@@ -131,32 +176,30 @@ abstract contract BridgeBase is OwnableUpgradeable, UUPSUpgradeable {
         if (state_with_proof.state.epoch != appliedEpoch + 1) {
             revert NotSuccessiveEpochs({epoch: appliedEpoch, nextEpoch: state_with_proof.state.epoch});
         }
-        uint256 common = (gasleftbefore - gasleft()) * tx.gasprice;
         uint256 statesLength = state_with_proof.state.states.length;
         uint256 stateHashIndex = 0;
         uint256 stateIndex = 0;
         while (stateIndex < statesLength) {
-            gasleftbefore = gasleft();
             while (stateHashIndex < state_with_proof.state_hashes.length) {
                 SharedStructs.ContractStateHash calldata state_hash = state_with_proof.state_hashes[stateHashIndex];
                 SharedStructs.StateWithAddress calldata state = state_with_proof.state.states[stateIndex];
                 if (localAddress[state_hash.contractAddress] == address(0)) {
                     unchecked {
                         ++stateHashIndex;
+                        ++stateIndex;
                     }
                     continue;
                 }
                 if (keccak256(state.state) != state_hash.stateHash) {
                     unchecked {
                         ++stateHashIndex;
+                        ++stateIndex;
                     }
                     continue;
                 }
-                uint256 used = (gasleftbefore - gasleft()) * tx.gasprice;
-                uint256 refund = (used + common / state_with_proof.state_hashes.length);
-                connectors[localAddress[state_hash.contractAddress]].applyStateWithRefund(
-                    state.state, payable(msg.sender), refund
-                );
+                if (isContract(address(connectors[localAddress[state_hash.contractAddress]]))) {
```

Comment:
Why?

Reply:
If it isn't a contract and we let it `applyState` I found during the test runs that it fails.